### PR TITLE
Fix incorrect memory accounting because of merges/mutations

### PR DESCRIPTION
**Pull request by azat**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in official stable or prestable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix incorrect memory accounting because of merges/mutations

In case of merges/mutations, separate MemoryTracker object is used, to
track memory separately, to report this metrics.

However, the tail of those allocations hadn't been taken into account,
and in case of lots of mutations/merges it can be significant.

100% reproducible with the following example:

Start the server w/o frequent sync with RSS:

    $ clickhouse-server -C /src/programs/server/config.xml -- --asynchronous_metrics_update_period_s=86400

Run the following snippet:

    create table data (key Int, value Nullable(Decimal(38, 0))) engine=ReplicatedMergeTree('/ch/data', '1') order by key
    insert into data values (1, -13028236692093846346337460743176821145)
    alter table data modify column value UInt256
    Code: 341. DB::Exception: Received from localhost:9000. DB::Exception: Exception happened during execution of mutation 'mutation_2.txt' with part 'all_1_1_0' reason: 'Code: 407. DB::Exception: Convert overflow: while executing 'FUNCTION _CAST(value :: 1, 'UInt256' :: 2) -> _CAST(value, 'UInt256') UInt256 : 3': (while reading from part ./store/dd9/dd9ef183-1270-4877-af22-8a000bdcc1b8/all_1_1_0/): While executing MergeTreeInOrder. (DECIMAL_OVERFLOW) (version 22.12.1.1)'. This error maybe retryable or not. In case of unretryable error, mutation can be killed with KILL MUTATION query. (UNFINISHED)

And track the memory:

    select value/1e9 from system.metrics where metric like 'MemoryTracking'

Reported-by: @filimonov
Follow-up for: #29691 (cc @nikitamikhaylov )

### CODE REVIEW

File: src/Storages/MergeTree/MergeList.cpp [hunk 1]
```diff
@@ -142,7 +142,11 @@ MergeInfo MergeListElement::getInfo() const
     return res;
 }
 
-MergeListElement::~MergeListElement() = default;
+MergeListElement::~MergeListElement()
+{
+    CurrentThread::getMemoryTracker()->adjustWithUntrackedMemory(untracked_memory);
+    untracked_memory = 0;
```
Comment (novikd) [inline src/Storages/MergeTree/MergeList.cpp:148]: Looks redundant

Comment (azat) [reply src/Storages/MergeTree/MergeList.cpp:148]: Yep, but just in case. Objections?

**Review by CheSema — APPROVED** [review]
LGTM

Seems like you found a bag.
- adjustWithUntrackedMemory should be call for untracked_memory at the end as it happens in ThreadStatus dtor.
- untracked_memory is signed by api, you fixed that.

**Comment by CheSema** [comment]
AST Fuzzer  -- https://github.com/ClickHouse/ClickHouse/issues/42819
Statefull -- https://github.com/ClickHouse/ClickHouse/pull/43696

**Comment by CheSema** [comment]
Integration tests -- https://github.com/ClickHouse/ClickHouse/issues/43795

**Pull request merged**